### ISSUE TITLE
inclusive_scan: std::int8_t input view and std::int32_t output view with CUDA backend produces incorrect result

### ISSUE THREAD

**Issue by cwsmith** [kokkos/kokkos#5513]
**Describe the bug**

I’m trying to compute a prefix sum with an `Kokkos::Experimental::inclusive_scan` on an array of std::int8_t and produce an array of std::int32_t and getting incorrect results with the CUDA backend.  The reproducer code is here:
https://github.com/cwsmith/kkReduce/blob/333a4fa9784dc44e57ee984b47e016b87a7823e7/kk_scan_shortToInt.cpp
The reproducer works with the OpenMP backend.

**Please also include the following items to support reproducing the bug**
1. compilers: 
   - host: GCC 10.1.0
   -  device: nvcc `Cuda compilation tools, release 11.4, V11.4.48, Build cuda_11.4.r11.4/compiler.30033411_0`  
2. Kokkos commit: develop@2ce06b9
3. platform and backend
   - OS: Red Hat Enterprise Linux Workstation release 7.9 (Maipo)
   - CPU: AMD Ryzen 9 3900X 12-Core Processor
   - GPU: NVIDIA GeForce RTX 2060 SUPER
4. cmake configure command

  - Kokkos
```
bkend=$1
d=buildKokkos${bkend}
omp=""
cuda=""
[[ ${bkend} == "OPENMP" ]] && omp="-DKokkos_ENABLE_OPENMP=on"
[[ ${bkend} == "CUDA" ]] && cuda="-DKokkos_ENABLE_CUDA=on -DKokkos_ENABLE_CUDA_LAMBDA=on"
cmake -S kokkos -B $d \
  -DCMAKE_CXX_COMPILER=g++ \
  -DKokkos_ARCH_TURING75=ON \
  -DBUILD_SHARED_LIBS=ON \
  $omp \
  $cuda \
  -DKokkos_ENABLE_SERIAL=ON \
  -DKokkos_ENABLE_DEBUG=on \
  -DKokkos_ENABLE_TESTS=off \
  -DCMAKE_INSTALL_PREFIX=$d/install
cmake --build $d -j 24 --target install
```
  - test code
  
```
bkend=$1
compiler=""
[[ ${bkend} == "OPENMP" ]] && compiler="-DCMAKE_CXX_COMPILER=g++"
[[ ${bkend} == "CUDA" ]] && compiler="-DCMAKE_CXX_COMPILER=$PWD/buildKokkos${bkend}/install/bin/nvcc_wrapper"
d=buildKkReduce${bkend}

cmake -S kkReduce -B $d \
$compiler \
-DKokkos_ROOT=buildKokkos${bkend}/install \
-DCMAKE_BUILD_TYPE=Debug

cmake --build $d --target scan_shortToInt
```

5. output from Kokkos cmake command
  - CUDA backend
```
-- Setting default Kokkos CXX standard to 17
-- The project name is: Kokkos
-- Using internal gtest for testing
-- Compiler Version: 11.4.48
-- kokkos_launch_compiler (/space/cwsmith/kokkosScan/kokkos/bin/kokkos_launch_compiler) is enabled...
-- Using -std=c++17 for C++17 standard as feature
-- Built-in Execution Spaces:
--     Device Parallel: Kokkos::Cuda
--     Host Parallel: NoTypeDefined
--       Host Serial: SERIAL
-- 
-- Architectures:
--  TURING75
-- Using internal desul_atomics copy
-- Kokkos Devices: CUDA;SERIAL, Kokkos Backends: CUDA;SERIAL
-- Configuring done
-- Generating done
-- Build files have been written to: /space/cwsmith/kokkosScan/buildKokkosCUDA
```
  - OpenMP backend
```
-- Setting default Kokkos CXX standard to 17
-- The CXX compiler identification is GNU 10.1.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/scorec/spack/v0154_2/install/linux-rhel7-x86_64/gcc-6.5.0/gcc-10.1.0-tf5jjaditemasrbsl7tz6pnqa6duqwkg/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting build type to 'RelWithDebInfo' as none was specified.
-- The project name is: Kokkos
-- Using internal gtest for testing
-- Using -std=gnu++17 for C++17 extensions as feature
-- Built-in Execution Spaces:
--     Device Parallel: NoTypeDefined
--     Host Parallel: Kokkos::OpenMP
--       Host Serial: SERIAL
-- 
-- Architectures:
-- Found TPLLIBDL: /usr/include  
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Using internal desul_atomics copy
-- Kokkos Devices: OPENMP;SERIAL, Kokkos Backends: OPENMP;SERIAL
-- Configuring done
-- Generating done
-- Build files have been written to: /space/cwsmith/kokkosScan/buildKokkosOPENMP
```
6. code needed to reproduce the bug: https://github.com/cwsmith/kkReduce/blob/333a4fa9784dc44e57ee984b47e016b87a7823e7/kk_scan_shortToInt.cpp
7. command line needed to reproduce the bug: `./kk_scan_shortToInt` (assuming the binary has the same name as the source, etc..) 
8. please also attach the `KokkosCore_config.h` header file (generated during the build)  [coreHeaders.zip](https://github.com/kokkos/kokkos/files/9699805/coreHeaders.zip)

**Any additional info**
none

**Comment by nmm0** [comment]
Thanks for letting us know about the issue, I was able to reproduce it. I'll work on a fix for it

**Comment by cwsmith** [comment]
Using develop@3c842fb, which includes #5555, the reproducer code runs correctly with the CUDA backend .  Thank you @dalg24.

**Comment by PhilMiller** [comment]
@nmm0 mentioned in #5545 that the change addressing that in #5555 would only partially fix the issue. Is there more left to be done, even though the given reproducer code now passes? Is there some variation on it that will still fail?

**Comment by PhilMiller** [comment]
Please make sure that concern doesn't get forgotten. Open another issue describing it, I guess?

**Comment by cwsmith** [comment]
For the record, two additional changes were needed in the reproducer (https://github.com/cwsmith/kkReduce/commit/71cddd0927fbba08a27e1d26bb09843516d0e136) to avoid overflows with an input array `std::int8_t` that had 128 or more entries.

**Comment by nmm0** [comment]
This issue was resolved in 4.0 and backported to 3.7.01. I'm going to close it. For the issue that @PhilMiller mentioned, I opened a new issue: https://github.com/kokkos/kokkos/issues/5985
